### PR TITLE
Add Faiss byte vector support blog

### LINKED ISSUE #3420: [BLOG] Introducing Byte Vector Support for Faiss in Opensearch Vector Engine
### Describe the blog post

The blog will explain the problem statement and describes how this Faiss Byte vector feature will help to reduce the memory requirements and cost besides improving the performance. Also, walks through how to use this feature as an end user and show case the improvements using the benchmarking results.

### Expected Title

Introducing Byte Vector Support for Faiss in Opensearch Vector Engine

### Authors Name

Naveen Tatikonda

### Authors Email

navtat@amazon.com

### Target Draft Date

11/04/2024

### Blog Post Category

technical

### Target Publication Date

11/11/2024

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by kolchfa-aws**
Add Faiss byte vector support blog

Closes #3420 

### Check List
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

**Comment by kolchfa-aws** [comment]
@pajuric Please edit the meta when you get a chance. Thank you!

**Review by naveentatikonda — APPROVED** [review]

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+|**Dataset 4**	|**1**	|2.91	|3.03	|3.16	|3.15	|
+|**Dataset 4**|**8**	|3.38	|3.30	|6.30	|4.75	|
+|**Dataset 4**	|**16**	|4.35	|3.80	|8.76	|8.83	|
+
+### Key findings
+
+When comparing the benchmarking results, here are the key findings:
+
+- **Memory savings**: Byte vectors reduced memory usage by up to **72%**, with higher-dimensional vectors achieving greater reductions.
+- **Indexing performance**: The mean indexing throughput for byte vectors was **2x to 107.84%** higher than for float vectors, especially with larger vector dimensions.
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:154]: "with larger vector dimensions" => "when using larger vector dimensions"?

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+- **Search performance**: Search latencies were similar, with byte vectors occasionally performing better.
+- **Recall**: For byte vectors, there was a slight (up to **8.8%**) reduction in recall compared to float vectors, depending on the dataset and the quantization technique used.
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:156]:
```suggestion
- **Recall**: For byte vectors, there was a slight (up to **8.8%**) reduction in recall as compared to float vectors, depending on the dataset and the quantization technique used.
```

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+## How does Faiss work with byte vectors internally?
+
+Faiss doesn't directly support the `byte` data type for storing vectors. To achieve this, OpenSearch uses a  [`QT_8bit_direct_signed` scalar quantizer](https://faiss.ai/cpp_api/struct/structfaiss_1_1ScalarQuantizer.html). This quantizer accepts float vectors within the signed 8-bit value range and encodes them as unsigned 8-bit integer vectors. During indexing and search, these encoded unsigned 8-bit integer vectors are decoded back into signed 8-bit original vectors for distance computation.
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:160]:
```suggestion
Faiss doesn't directly support the `byte` data type for vector storage. To achieve this, OpenSearch uses a  [`QT_8bit_direct_signed` scalar quantizer](https://faiss.ai/cpp_api/struct/structfaiss_1_1ScalarQuantizer.html). This quantizer accepts float vectors within the signed 8-bit value range and encodes them as unsigned 8-bit integer vectors. During indexing and search, these encoded unsigned 8-bit integer vectors are decoded back into signed 8-bit original vectors for distance computation.
```

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:161]: Line 160, end of last sentence: "are decoded back into the original signed 8-bit vectors for the purpose of distance computation"?

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+This quantization approach reduces the memory footprint by a factor of four. However, encoding and decoding during scalar quantization introduce additional latency. To mitigate this, you can use [SIMD optimization](https://opensearch.org/docs/latest/search-plugins/knn/knn-index#simd-optimization-for-the-faiss-engine) with the `QT_8bit_direct_signed` quantizer to reduce search latencies and improve indexing throughput.
+
+### Example
+
+The following example shows how an input vector is encoded and decoded using `QT_8bit_direct_signed` scalar quantizer:
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:166]:
```suggestion
The following example shows how an input vector is encoded and decoded using the `QT_8bit_direct_signed` scalar quantizer:
```

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+// Encoded vector generated by adding 128 to each dimension of the input vector to convert signed int8 to unsigned int8:
+[2, 156, 255, 128, 138, 83, 140, 18]
+
+// Encoded vector is decoded back into the original signed int8 vector by subtracting 128 from each dimension for distance computation:
+[-126, 28, 127, 0, 10, -45, 12, -110]
+```
+
+## Conclusion
+
+OpenSearch 2.17 introduced support for Faiss byte vectors, enabling you to store quantized byte vector embeddings efficiently. This reduces memory consumption by up to 75%, lowers costs, and maintains high performance. These advantages make byte vectors an excellent choice for large-scale similarity search applications, especially where memory resources are limited, and applications that handle large volumes of data within the signed byte value range.
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:181]:
```suggestion
OpenSearch 2.17 introduced support for Faiss byte vectors, allowing you to efficiently store quantized byte vector embeddings. This reduces memory consumption by up to 75%, lowers costs, and maintains high performance. These advantages make byte vectors an excellent choice for large-scale similarity search applications, especially when memory resources are limited, and applications that handle large volumes of data within the signed byte value range.
```

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+## Future enhancements
+
+In future versions, we plan to enhance this feature by adding an `on_disk` mode with a `4x` compression level in Faiss. This mode will accept `fp32` vectors as input, perform online training, and quantize the data into byte-sized vectors, eliminating the need for performing external quantization.
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:185]:
```suggestion
In future versions, we plan to enhance this feature by adding an `on_disk` mode with a `4x` compression level in Faiss. This mode will accept `fp32` vectors as input, perform online training, and quantize the data into byte-sized vectors, eliminating the need to perform external quantization.
```

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:186]: Line 185, first sentence: "compression level in Faiss" => "Faiss compression level"?

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+// Encoded vector is decoded back into the original signed int8 vector by subtracting 128 from each dimension for distance computation:
+[-126, 28, 127, 0, 10, -45, 12, -110]
+```
+
+## Conclusion
+
+OpenSearch 2.17 introduced support for Faiss byte vectors, enabling you to store quantized byte vector embeddings efficiently. This reduces memory consumption by up to 75%, lowers costs, and maintains high performance. These advantages make byte vectors an excellent choice for large-scale similarity search applications, especially where memory resources are limited, and applications that handle large volumes of data within the signed byte value range.
+
+## Future enhancements
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:183]: Nit: This section should probably precede the conclusion.

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Editorial review complete. Please see my comments and changes and let me know if you have any questions. Thanks!

Cc: @pajuric

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+
+Using byte vectors instead of float vectors for vector search provides significant improvements in memory efficiency and performance. This is especially beneficial for large-scale vector databases or environments with limited resources. Faiss byte vectors enable you to store quantized embeddings, significantly reducing memory consumption and lowering costs. This approach typically results in only minimal recall loss compared to using full-precision (float) vectors.
+
+
+## How to use a Faiss byte vector?
+
+A byte vector is a compact representation where each dimension is a signed 8-bit integer ranging from -128 to 127. To use byte vectors, you must convert your input vectors, typically in `float` format, into the `byte` type before ingestion. This process requires quantization techniques, which compress float vectors while maintaining essential data characteristics. For more information, see [Quantization techniques](https://opensearch.org/docs/latest/field-types/supported-field-types/knn-vector#quantization-techniques).
```
Comment (kolchfa-aws) [inline _posts/2024-11-22-faiss-byte-vector.md:25]:
```suggestion
A byte vector is a compact vector representation in which each dimension is a signed 8-bit integer ranging from -128 to 127. To use byte vectors, you must convert your input vectors, typically in `float` format, into the `byte` type before ingestion. This process requires quantization techniques, which compress float vectors while maintaining essential data characteristics. For more information, see [Quantization techniques](https://opensearch.org/docs/latest/field-types/supported-field-types/knn-vector#quantization-techniques).
```

[the remark this answers is not in the document]
Comment (kolchfa-aws) [reply _posts/2024-11-22-faiss-byte-vector.md:25]: Reworded.

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+The following table lists the cluster configuration for benchmarking tests.
+
+|`m`	|`ef_construction`	|`ef_search`	|Replicas	|Primary shards	|Indexing clients	|
+|---	|---	|---	|---	|---	|---	|
+|16	|100	|100	|0	|8	|16	|
+
+The following table lists the dataset configuration for benchmarking tests.
+
+|Dataset ID	|Dataset	|Dimension of vector	|Data size	|Number of queries	|Training data range	|Query data range	|Space type	|
```
Comment (kolchfa-aws) [inline _posts/2024-11-22-faiss-byte-vector.md:113]:
```suggestion
|Dataset ID	|Dataset	|Vector dimension	|Data size	|Number of queries	|Training data range	|Query data range	|Space type	|
```

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+|---	|---	|---	|---	|---	|---	|---	|---	|
+|**Dataset 1**	|gist-960-euclidean	|960	|1,000,000	|1,000	|[0.0, 1.48]	|[0.0, 0.729]	|L2	|
+|**Dataset 2**	|cohere-ip-10m	|768	|10,000,000	|10,000	|[-4.142334, 5.5211477]	|[-4.109505, 5.4809895]	|innerproduct	|
+|**Dataset 3**	|cohere-ip-1m	|768	|1,000,000	|10,000	|[-4.1073565, 5.504557]	|[-4.109505, 5.4809895]	|innerproduct	|
+|**Dataset 4**	|sift-128-euclidean	|128	|1,000,000	|10,000	|[0.0, 218.0]	|[0.0, 184.0]	|L2	|
+
+### Recall, memory, and indexing results
+
+|Dataset ID	|Faiss HNSW recall@100	|Faiss HNSW-byte recall@100	|% Reduction in recall	|Faiss HNSW memory usage (GB)	|Faiss HNSW byte memory usage (GB)	|% Reduction in memory	|Faiss HNSW mean indexing throughput (docs/sec)	|Faiss HNSW byte mean indexing throughput (docs/sec)	|% Gain in indexing throughput	|
```
Comment (kolchfa-aws) [inline _posts/2024-11-22-faiss-byte-vector.md:122]:
```suggestion
|Dataset ID	|Faiss HNSW recall@100	|Faiss HNSW byte recall@100	|% Reduction in recall	|Faiss HNSW memory usage (GB)	|Faiss HNSW byte memory usage (GB)	|% Reduction in memory	|Faiss HNSW mean indexing throughput (docs/sec)	|Faiss HNSW byte mean indexing throughput (docs/sec)	|% Gain in indexing throughput	|
```

[the remark this answers is not in the document]
Comment (kolchfa-aws) [reply _posts/2024-11-22-faiss-byte-vector.md:122]: Removed

File: _posts/2024-11-22-faiss-byte-vector.md (hunk 4 above)
Comment (kolchfa-aws) [reply _posts/2024-11-22-faiss-byte-vector.md:161]: I think it's understandable in the original form.

Comment (natebower) [reply _posts/2024-11-22-faiss-byte-vector.md:161]: It reads as though "the" should precede "signed", though.

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+**Note**: When using `byte` vectors, expect some loss of precision in the recall compared to using `float` vectors. Byte vectors are useful in large-scale applications and use cases that prioritize reducing memory usage in exchange for a minimal loss in recall.
+
+
+## Benchmarking results
+
+We ran benchmarking tests on popular datasets using OpenSearch Benchmark to compare recall, indexing, and search performance between float vectors and byte vectors using Faiss HNSW.
+
+**Note**: Without SIMD optimization (such as AVX2 or NEON) or when AVX2 is disabled (on x86 architectures), the quantization process introduces additional latency. For details on AVX2-compatible processors, see [CPUs with AVX2](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions#CPUs_with_AVX2). In an AWS environment, all community Amazon Machine Images (AMIs) with [HVM](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/virtualization_types.html) support AVX2 optimization.
+
```
[the remark this answers is not in the document]
Comment (kolchfa-aws) [reply _posts/2024-11-22-faiss-byte-vector.md:100]: Done

File: _posts/2024-11-22-faiss-byte-vector.md (hunk 1 above)
Comment (kolchfa-aws) [reply _posts/2024-11-22-faiss-byte-vector.md:154]: I think it's ok as is

File: _posts/2024-11-22-faiss-byte-vector.md (hunk 9 above)
Comment (kolchfa-aws) [reply _posts/2024-11-22-faiss-byte-vector.md]: Done

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+- **Memory savings**: Byte vectors reduced memory usage by up to **72%**, with higher-dimensional vectors achieving greater reductions.
+- **Indexing performance**: The mean indexing throughput for byte vectors was **2x to 107.84%** higher than for float vectors, especially with larger vector dimensions.
+- **Search performance**: Search latencies were similar, with byte vectors occasionally performing better.
+- **Recall**: For byte vectors, there was a slight (up to **8.8%**) reduction in recall as compared to float vectors, depending on the dataset and the quantization technique used.
+
+## How does Faiss work with byte vectors internally?
+
+Faiss doesn't directly support the `byte` data type for vector storage. To achieve this, OpenSearch uses a  [`QT_8bit_direct_signed` scalar quantizer](https://faiss.ai/cpp_api/struct/structfaiss_1_1ScalarQuantizer.html). This quantizer accepts float vectors within the signed 8-bit value range and encodes them as unsigned 8-bit integer vectors. During indexing and search, these encoded unsigned 8-bit integer vectors are decoded back into signed 8-bit original vectors for distance computation.
```
Comment (kolchfa-aws) [inline _posts/2024-11-22-faiss-byte-vector.md:160]:
```suggestion
Faiss doesn't directly support the `byte` data type for vector storage. To achieve this, OpenSearch uses a  [`QT_8bit_direct_signed` scalar quantizer](https://faiss.ai/cpp_api/struct/structfaiss_1_1ScalarQuantizer.html). This quantizer accepts float vectors within the signed 8-bit value range and encodes them as unsigned 8-bit integer vectors. During indexing and search, these encoded unsigned 8-bit integer vectors are decoded back into the original signed 8-bit vectors for distance computation.
```

File: _posts/2024-11-22-faiss-byte-vector.md [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+```
+
+**Note**: When using `byte` vectors, expect some loss of recall precision as compared to using `float` vectors. Byte vectors are useful for large-scale applications and use cases that prioritize reducing memory usage in exchange for a minimal loss in recall.
+
+
+## Benchmarking results
+
+We used OpenSearch Benchmark to run benchmarking tests on popular datasets to compare recall, indexing, and search performance between float vectors and byte vectors using Faiss HNSW.
+
+**Note**: Without Single Instruction, Multiple Data (SIMD) optimization (such as AVX2 or NEON) or when AVX2 is disabled (on x86 architectures), the quantization process introduces additional latency. For more information about AVX2-compatible processors, see [CPUs with AVX2](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions#CPUs_with_AVX2). In an AWS environment, all community Amazon Machine Images (AMIs) with [HVM](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/virtualization_types.html) support AVX2 optimization.
```
Comment (natebower) [inline _posts/2024-11-22-faiss-byte-vector.md:99]:
```suggestion
**Note**: Without single instruction, multiple data (SIMD) optimization (such as AVX2 or NEON) or when AVX2 is disabled (on x86 architectures), the quantization process introduces additional latency. For more information about AVX2-compatible processors, see [CPUs with AVX2](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions#CPUs_with_AVX2). In an AWS environment, all community Amazon Machine Images (AMIs) with [HVM](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/virtualization_types.html) support AVX2 optimization.
```

File: _posts/2024-11-22-faiss-byte-vector.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (natebower) [reply _posts/2024-11-22-faiss-byte-vector.md:100]: Changed to lowercase

**Comment by kolchfa-aws** [comment]
@pajuric Editorial comments addressed. This blog will be ready to publish once we update the meta and date. Thanks!

**Review by nateynateynate — APPROVED** [review]

**Pull request merged**